### PR TITLE
Refactoring of the sample:

### PR DESCRIPTION
**Pull request by jmprieur**
Refactoring of the sample to:
- enable better reusability
- highlight the code that does the magic.
- improve performance by reusing the HttpClient.

### CODE REVIEW

File: device-code-flow-console/MyInformation.cs [hunk 1]
```diff
@@ -32,185 +31,46 @@
...
                 Console.WriteLine();
 
                 Console.ForegroundColor = ConsoleColor.White;
                 Console.WriteLine("My manager");
                 Console.ResetColor();
-                await CallWebApiAndDisplayResultASync(WebApiUrlMyManager, authenticationResult);
-            }
-        }
-
-        /// <summary>
```
Comment (jmprieur) [inline device-code-flow-console/MyInformation.cs]: This method moved to the base class

File: device-code-flow-console/MyInformation.cs [hunk 2]
```diff
@@ -32,185 +31,46 @@
...
         /// </summary>
         /// <returns></returns>
         public async Task DisplayMeAndMyManagerAsync()
         {
-            AuthenticationConfig config = AuthenticationConfig.ReadFromJsonFile("appsettings.json");
-
-            app = new PublicClientApplication(config.ClientId, config.Authority);
-
-            AuthenticationResult authenticationResult = await AcquireATokenFromCacheOrDeviceCodeFlow();
+            AuthenticationResult authenticationResult = await AcquireATokenFromCacheOrDeviceCodeFlow(Scopes);
```
Comment (bgavrilMS) [inline device-code-flow-console/MyInformation.cs]: This method does too much, please create a private method for displaying the graph stuff, something like CallGraphAndDisplayResults(string authToken)

File: device-code-flow-console/Program.cs [hunk 3]
```diff
@@ -53,5 +55,15 @@ static void Main(string[] args)
                 Console.ResetColor();
             }
         }
+
+        private static async Task RunAsync()
+        {
+            AuthenticationConfig config = AuthenticationConfig.ReadFromJsonFile("appsettings.json");
+            var app = new PublicClientApplication(config.ClientId, config.Authority);
+            var httpClient = new HttpClient();
```
Comment (bgavrilMS) [inline device-code-flow-console/Program.cs]: Do you have a reason to inject the httpClient into MyInformation ? I know it is good practice to use a single instance of HttpClient in the app, but MyInformation is the only component that uses it, so might as well let it create it.

File: device-code-flow-console/MyInformation.cs [hunk 4]
```diff
@@ -32,185 +31,46 @@
 
 namespace device_code_flow_console
 {
-    public class MyInformation
+    public class MyInformation : PublicAppUsingDeviceCodeFlow
```
Comment (bgavrilMS) [inline device-code-flow-console/MyInformation.cs]: This inheritance is not natural. I would use composition instead. Makes for much more readable code.

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        /// For more information see https://aka.ms/msal-net-device-code-flow
+        /// </remarks>
+        protected PublicAppUsingDeviceCodeFlow(PublicClientApplication app, HttpClient httpClient)
+        {
+            App = app;
+            Client = httpClient;
+        }
+        protected PublicClientApplication App { get; private set; }
+
+        protected HttpClient Client { get; private set; }
```
Comment (bgavrilMS) [inline device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: HttpClient

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                    result = await App.AcquireTokenSilentAsync(scopes, accounts.FirstOrDefault());
+                }
+                catch (MsalUiRequiredException)
+                {
+                    result = null;
+                }
+            }
+            else
+            {
+                result = null;
```
Comment (bgavrilMS) [inline device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: initialize result with null and eliminate this else clause

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                // See https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads 
+                // for more detailed information on how C# supports cancellation in managed threads.
+                result = null;
+            }
+
+            catch (MsalClientException ex)
+            {
+                string errorCode = ex.ErrorCode;
+
+                // Verification code expired before contacting the server
```
Comment (bgavrilMS) [inline device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: This is missing the error code for this particular error.

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            return result;
+        }
+
+        /// <summary>
+        /// Calls the protected Web API and processes the result
+        /// </summary>
+        /// <param name="webApiUrl">Url of the Web API to call (supposed to return Json)</param>
+        /// <param name="accessToken">Access token used as a bearer security token to call the Web API</param>
+        /// <param name="processResult">Callback used to process the result of the call to the Web API</param>
+        public async Task CallWebApiAndDisplayResultASync(string webApiUrl, string accessToken, Action<JObject> processResult)
```
Comment (bgavrilMS) [inline device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: If you are going to go with composition, I would move this method to MyInformation class

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs (hunk 7 above)
Comment (jmprieur) [reply device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: yes, it does not seem to be present in MsalClientException?

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs (hunk 8 above)
Comment (jmprieur) [reply device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: Thanks. I've done it

File: device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs (hunk 6 above)
Comment (jmprieur) [reply device-code-flow-console/PublicAppUsingDeviceCodeFlow.cs]: Done. Thanks!

File: device-code-flow-console/MyInformation.cs (hunk 4 above)
Comment (jmprieur) [reply device-code-flow-console/MyInformation.cs]: Done. Thanks!

File: device-code-flow-console/MyInformation.cs (hunk 2 above)
Comment (jmprieur) [reply device-code-flow-console/MyInformation.cs]: done. Thanks!

**Pull request merged**